pkg/tfbridge/provider.go

```diff
@@ -678,6 +679,39 @@ func (p *Provider) Check(ctx context.Context, req *pulumirpc.CheckRequest) (*pul
 	return &pulumirpc.CheckResponse{Inputs: minputs, Failures: failures}, nil
 }
 
+// For properties with MaxItemsOne, where the state is still an array
+// (i.e. from a previous version without MaxItemsOne)
+// we need to mark them for update manually in order to correct the state
+// from an array to a flat type.
+func markWronglyTypedStateDiff(res Resource, olds resource.PropertyMap, diff *shim.InstanceDiff) error {
+	if diff == nil {
+		return fmt.Errorf("diff is nil")
+	}
+	detectMismatchingType := func(localSchema shim.Schema, localInfo *SchemaInfo, p resource.PropertyValue) bool {
+		if localInfo != nil &&
+			localInfo.MaxItemsOne != nil &&
+			*localInfo.MaxItemsOne &&
+			localInfo.Elem == nil &&
+			p.IsArray() {
+			return true
```

Comment:
I think it's worth commenting/ naming this function better that it's very surgical. Not all type mismatches are detected but specifically the ones originating from a MaxItemsOne mismatch. Worth commenting also that we expect MakeTerraformInputs to automatically correct the state data to the expected shape.